metagraph/core/compiler.py

```diff
@@ -29,64 +30,108 @@ def extract_compilable_subgraphs(
+        )
+
+    for next_key in ordered_keys:
+        next_key_dependencies = dependencies[next_key]
+        key_dependents = dependents[key]
 
         if (
-            isinstance(task_callable, DelayedAlgo)
-            and task_callable.algo._compiler == compiler
+            len(next_key_dependencies) == 1
+            and len(key_dependents) == 1
+            and next_key in key_dependents
+            and key in next_key_dependencies
         ):
-            compilable_keys.add(key)
+            current_chain.append(next_key)
+        elif len(current_chain) >= chain_threshold:
+            _note_subgraph(current_chain)
+            current_chain = [next_key]
+        else:
+            current_chain = [next_key]
+        key = next_key
+
+    if len(current_chain) >= chain_threshold:
+        _note_subgraph(current_chain)
+
+    return subgraphs
+
+
+def _get_compilable_dask_keys(dsk: Dict, compiler: str) -> Tuple[set, set]:
+
+    compilable_keys = set()
+    for key in dsk.keys():
+        task_callable = dsk[key][0]
+        if isinstance(task_callable, DelayedAlgo):
+            if task_callable.algo._compiler == compiler:
+                compilable_keys.add(key)
 
     non_compilable_keys = set(dsk.keys()) - compilable_keys
 
-    subgraphs = []
-    # keep trying to build chains until no compilable tasks are left
-    keys_left = compilable_keys.copy()
-    while len(keys_left) > 0:
-        key = keys_left.pop()
-        chain = [key]
-        chain_extended = True
-        while chain_extended:
-            next_compilable_tasks = dependents[chain[-1]] & compilable_keys
-            prior_compilable_tasks = dependencies[chain[0]] & compilable_keys
-
-            # Can we extend the chain?
-            chain_extended = False
-
-            if len(next_compilable_tasks) == 1:
-                candidate = next_compilable_tasks.pop()
-                # ensure this task does not have multiple compileable dependencies
-                if len(dependencies[candidate] & compilable_keys) == 1:
-                    chain.append(candidate)
-                    keys_left.remove(chain[-1])
-                    chain_extended = True
-
-            if len(prior_compilable_tasks) == 1:
-                chain.insert(0, prior_compilable_tasks.pop())
-                keys_left.remove(chain[0])
-                chain_extended = True
-
-        if len(chain) >= chain_threshold:
-            # collect all the inputs to the tasks in this chain
-            inputs = set()
-            chain_keys = set(chain)
-            for key in chain:
-                inputs |= dependencies[key] - chain_keys
-            tasks = {key: dsk[key] for key in chain}
-            subgraphs.append(
-                DaskSubgraph(tasks=tasks, input_keys=list(inputs), output_key=chain[-1])
-            )
+    return compilable_keys, non_compilable_keys
 
-    return subgraphs
+
+def _topologically_sorted_dask_keys(
```

Comment:
On second thought, it's not clear to me that a topological sort is needed here. Using DFS might suffice.

Reply:
Can we call Dask's toposort implementation?

top level function is here:

https://github.com/dask/dask/blob/master/dask/core.py#L429

but the real implementation is here:

https://github.com/dask/dask/blob/master/dask/core.py#L360

Reply:
https://github.com/metagraph-dev/metagraph/pull/166/commits/2f64b0cc92fb43817a1a6298c8567a98ddbbc787 makes it so that we use Dask's toposort implementation.

Reply:
@seibert , it turns out that we didn't need a topological sort. We just need a DFS ordering. My topological sort implementation worked since it yielded nodes in a DFS-like order, e.g. for this graph,
```
               a
              / \
             b   d
             |   |
             c   e
```
my implementation might return `[a,d,e,b,c]` but will never return `[a,b,d,c,e]` even though the latter is in correct topological order. Dask's topological sort might return the latter, which was causing some of the tests to fail. 


https://github.com/metagraph-dev/metagraph/pull/166/commits/068cca3967bdc7594575a8d6c0efbeee30c43e30 makes `extract_compilable_subgraphs` use DFS.